crates/ra_assists/src/move_guard_to_arm_body_and_back.rs

```diff
@@ -0,0 +1,264 @@
+use hir::db::HirDatabase;
+use ra_syntax::{
+    TextUnit,
+    SyntaxElement,
+    ast::{MatchArm, AstNode, AstToken, IfExpr},
+    ast,
+};
+
+use crate::{AssistCtx, Assist, AssistId};
+
+pub(crate) fn move_guard_to_arm_body(mut ctx: AssistCtx<impl HirDatabase>) -> Option<Assist> {
+    let match_arm = ctx.node_at_offset::<MatchArm>()?;
+    let guard = match_arm.guard()?;
+    let space_before_guard = guard.syntax().prev_sibling_or_token();
+
+    let guard_conditions = guard.expr()?;
+    let arm_expr = match_arm.expr()?;
+    let buf = format!("if {} {{ {} }}", guard_conditions.syntax().text(), arm_expr.syntax().text());
+
+    ctx.add_action(AssistId("move_guard_to_arm_body"), "move guard to arm body", |edit| {
+        edit.target(guard.syntax().range());
+        let offseting_amount = match space_before_guard {
+            Some(SyntaxElement::Token(tok)) => {
+                if let Some(_) = ast::Whitespace::cast(tok) {
+                    let ele = space_before_guard.unwrap().range();
+                    edit.delete(ele);
+                    ele.len()
+                } else {
+                    TextUnit::from(0)
+                }
+            }
+            _ => TextUnit::from(0),
+        };
+
+        edit.delete(guard.syntax().range());
+        edit.replace_node_and_indent(arm_expr.syntax(), buf);
+        edit.set_cursor(arm_expr.syntax().range().start() + TextUnit::from(3) - offseting_amount);
+    });
+    ctx.build()
+}
+
+pub(crate) fn move_arm_cond_to_match_guard(mut ctx: AssistCtx<impl HirDatabase>) -> Option<Assist> {
+    let match_arm: &MatchArm = ctx.node_at_offset::<MatchArm>()?;
+    let last_match_pat = match_arm.pats().last()?;
+
+    let arm_body = match_arm.expr()?;
+    let if_expr: &IfExpr = IfExpr::cast(arm_body.syntax())?;
+    let cond = if_expr.condition()?;
+    let then_block = if_expr.then_branch()?;
+
+    // Not support if with else branch
+    if let Some(_) = if_expr.else_branch() {
+        return None;
+    }
+
+    let buf = format!(" if {}", cond.syntax().text());
+
+    ctx.add_action(
+        AssistId("move_arm_cond_to_match_guard"),
+        "move condition to match guard",
+        |edit| {
+            edit.target(last_match_pat.syntax().range());
```

Comment:
Let's set the `if_expr` as target. Target should be some node, containg the cursor